### PR TITLE
Allow you to skip the factory keyword.

### LINKED ISSUE #168: Make `clients.acme.factory` optional and use discovery instead
| Q            | A
| ------------ | ---
| Bug?         | no
| New Feature? | yes
| Version      | N/A

I propose to make the `factory` key of the client configuration optional (`clients.acme.factory`).
In such case, we will use the discovery to find an available implementation. Of course, the `config` node would be disabled when no factory is set.

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #168

### CODE REVIEW

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -173,9 +173,15 @@ private function configureClients(ArrayNodeDefinition $root)
                     })
                     ->thenInvalid('A http client can\'t be decorated with several of FlexibleHttpClient, HttpMethodsClient and BatchClient. Only one of the following options can be true. ("flexible_client", "http_methods_client", "batch_client")')
                 ->end()
+                ->validate()
+                    ->ifTrue(function ($config) {
+                        return $config['factory'] === 'httplug.factory.auto' && !empty($config['config']);
+                    })
+                    ->thenInvalid('You must specify a valid "factory" in order to use the "config"')
```
Comment (xabbuh) [inline DependencyInjection/Configuration.php]: IMO this message is not really clear. What does "valid config" mean in this context?

File: ClientFactory/AutoDiscoveryFactory.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class AutoDiscoveryFactory implements ClientFactory
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function createClient(array $config = [])
+    {
+        return HttpClientDiscovery::find();
```
Comment (fbourigault) [inline ClientFactory/AutoDiscoveryFactory.php]: Does it worth to fallback to `HttpAsyncClientDiscovery` if `HttpClientDiscovery` does not returned any implementation?

File: DependencyInjection/Configuration.php (hunk 1 above)
Comment (Nyholm) [inline DependencyInjection/Configuration.php]: Thank you. Is it better now?

File: ClientFactory/AutoDiscoveryFactory.php (hunk 2 above)
Comment (Nyholm) [inline ClientFactory/AutoDiscoveryFactory.php]: I doubt it. I believe it is very rare that an async adapter/client does not implement HttpClient interface (not async).

File: DependencyInjection/Configuration.php (hunk 1 above)
Comment (xabbuh) [inline DependencyInjection/Configuration.php]: Looks better

**Comment by Nyholm** [comment]
Thank you for the reviews.

**Pull request merged**